bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java

```diff
@@ -262,10 +278,30 @@ public void seal(SegmentHandle handle) throws StreamSegmentException {
     public void unseal(SegmentHandle handle) throws StreamSegmentException {
         ensureInitializedAndNotClosed();
         long traceId = LoggerHelpers.traceEnter(log, "seal", handle);
+        Path targetPath = getFilePath(handle.getSegmentName(), this.epoch);
+        Path sealedPath = getSealedFilePath(handle.getSegmentName());
+
         try {
-            FileStatus status = findStatusForSegment(handle.getSegmentName(), true);
-            makeWrite(status);
-            this.fileSystem.rename(status.getPath(), getFilePath(handle.getSegmentName(), this.epoch));
+            makeWrite(sealedPath);
+            this.fileSystem.rename(sealedPath, targetPath);
+        } catch (PathNotFoundException | FileNotFoundException e) {
+            try {
+                FileStatus status = findStatusForSegment(handle.getSegmentName(), true);
+                if (isSealed(status.getPath())) {
+                    makeWrite(status.getPath());
```

Comment:
Same here

Reply:
For the operation seal I'm using rename before making file read-only. To possibly leaving file read only when seal fails without sealing it.
For unseal , I'm using opposite sequence.